dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,374 @@
+ *
+ * For details on the way the convolution is computed, see the docs of the actual cuda kernel.
+ */
+template <typename Out, typename In, typename W, bool has_channel_dim, bool has_sequence_dim,
+          int axes_, bool enable_roi_>
+struct FilterGpu {
+  /* It computes a correlation of the input and the filter.
+  Flip filter in both dimensions for a convolution. */
+
+  static constexpr int axes = axes_;
+  static constexpr bool enable_roi = enable_roi_;
+  static constexpr int ndim =
+      static_cast<int>(has_sequence_dim) + axes + static_cast<int>(has_channel_dim);
+  static constexpr int sequence_dim = has_sequence_dim ? 0 : -1;
+  static constexpr int channels_dim = has_channel_dim ? ndim - 1 : -1;
+  static_assert(axes == 2 || axes == 3);
+  using Intermediate = decltype(std::declval<W>() * std::declval<In>());
+  using StaticConfigT = filter::StaticConfig<axes>;
+  using BlockSetupProviderT = filter::AdaptiveBlock<StaticConfigT>;
+  using BlockSetupT = typename BlockSetupProviderT::BlockSetup;
+  using StaticBlockFactoryT = filter::StaticBlock<StaticConfigT>;
+  using StaticBlockT = typename StaticBlockFactoryT::BlockSetup;
+  using GridSetupT = filter::GridSetup<StaticConfigT>;
+  using InShapeDescT = filter::InShapeDesc<axes>;
+  using OutShapeDescT = filter::OutShapeDesc<axes>;
+  using WorkspaceDescT = filter::WorkspaceDesc<axes>;
+  using SampleDescT = filter::SampleDesc<Out, In, W, Intermediate, axes>;
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim>& out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageGPU, const W, axes>& filters,
+           const span<const ivec<axes>> anchors, boundary::BoundaryType border_type,
+           const TensorListView<StorageGPU, const In, 0>& fill_values = {}) {
+    auto num_samples = in.shape.num_samples();
+    assert(out.num_samples() == num_samples && filters.num_samples() == num_samples &&
+           anchors.size() == num_samples);
+    assert(fill_values.num_samples() == num_samples || fill_values.num_samples() == 0);
+
+    SetupSampleDescs(out, in, filters, anchors);
+    SampleDescT* samples_desc_dev;
+    std::tie(samples_desc_dev) = ctx.scratchpad->ToContiguousGPU(ctx.gpu.stream, samples_desc_);
+    auto grid_setup = PrepareGridSetup(make_cspan(samples_desc_));
+    WithBlockSetupProvider(ctx, [&](auto&& block_setup_provider) {
+      WithInLoaderProvider(ctx, border_type, fill_values, [&](auto&& in_loader_provider) {
+        WithOutShapeProvider([&](auto&& out_shape_provider) {
+          WithConvFactory([&](auto&& conv_factory) {
+            filter::filter<<<grid_setup.kernel_setup(), StaticConfigT::threadblock_size,
+                             required_shm_size_, ctx.gpu.stream>>>(
+                samples_desc_dev, block_setup_provider, in_loader_provider, grid_setup,
+                out_shape_provider, conv_factory);
+            CUDA_CALL(cudaGetLastError());
```

Comment:
Well, I guess it would be better to split the launch instead of redirecting all sample to the inefficient global load use case. This can be left for a follow-up, though.